src/rep.rs

```diff
@@ -75,51 +60,28 @@ impl Socket for RepSocket {
     }
 }
 
-#[async_trait]
 impl MultiPeerBackend for RepSocketBackend {
-    async fn peer_connected(
-        &self,
-        peer_id: &PeerIdentity,
-    ) -> (mpsc::Receiver<Message>, oneshot::Receiver<bool>) {
-        let default_queue_size = 100;
-        let (out_queue, out_queue_receiver) = mpsc::channel(default_queue_size);
-        let (in_queue, in_queue_receiver) = mpsc::channel::<Message>(default_queue_size);
-        let (stop_handle, stop_callback) = oneshot::channel::<bool>();
+    fn peer_connected(self: Arc<Self>, peer_id: &PeerIdentity, io: FramedIo) {
+        let (recv_queue, send_queue) = io.into_parts();
 
         self.peers.insert(
             peer_id.clone(),
             RepPeer {
                 _identity: peer_id.clone(),
-                send_queue: out_queue,
-                recv_queue_in: in_queue,
-                _io_close_handle: stop_handle,
+                send_queue,
             },
         );
-        self.peer_queue_in
-            .clone()
-            .try_send((peer_id.clone(), in_queue_receiver))
-            .unwrap();
-
-        (out_queue_receiver, stop_callback)
+        self.fair_queue_inner
+            .lock()
```

Comment:
Curious about the ramifications of this - I thought that since acquiring a mutex is a blocking operation, doing it in async code would result in all tasks on the current thread in the async executor being blocked? Is this actually performant when scaling to many async tasks at once? Perhaps its faster on one thread, but I am suspicious of how this scales

Reply:
Good question. In theory it may become an issue. However my thoughts are currently the following:
1) Contention happens always between 2 threads (one processing incoming connections, and the other is user thread actually receiving data). So contention doesn't become worse with more spawned tasks. I assume that in most cases it would just take an atomic operation or perhaps a spin lock to actually take this mutex.
2) Changes made by this MR were made after some benchmarking and comparing with C code doing equivalent stuff. After this changes performance improved and now it's even a bit better than C++ implementation. If you could provide some kind of test scenario when this becomes an issue we can think what we can do about it.
3) I didn't find a better way of doing this. Basically 2 threads need to communicate and share data. Sharing data via channels also has some internal synchronization. If you can propose a better way of doing this we can try to benchmark and discuss it.

Reply:
What about using the async version? https://docs.rs/tokio/1.0.0/tokio/sync/struct.Mutex.html
I could submit a PR that switched to the async mutex and we can benchmark to see if there is a performance regression. Barring a regression, we could then switch to the async version "just in case"

Reply:
I've seen this mutex already, basically it's description says:
>The primary use case of the async mutex is to provide shared mutable access to IO resources such as a database connection. If the data stored behind the mutex is just data, it is often better to use a blocking mutex such as the one in the standard library or parking_lot. This is because the feature that the async mutex offers over the blocking mutex is that it is possible to keep the mutex locked across an .await point, which is rarely necessary for data.

I don't need to carry locked mutex over await so I decided that simple sync mutex should be sufficient.

>Can you elaborate? I didn't fully understand this

Efficiency and performance of a single mutex is determined by 2 main factors:
1) Length of critical sections (time while the mutex is locked). I tried to minimize this by using only a simplest operations while the mutex is locked.
2) Total number of threads trying to lock mutex concurrently. I was actually wrong claiming that only 2 thread content for the mutex (actually there are 3 of them - 1 thread processing incoming connections, 2 user thread, 3 async reactor thread that notifies future when new data received). Events generated by thread 1 (the code you are pointing to) only happen occasuly so it should not affect performance a lot. Main contention happens between thread 2 and 3.

From what I know about the implementation of parking lot mutex (and some other mutexes as well) It can work in 3 basic modes:
1) trying to lock an unlocked mutex - this happens very fast cause it's done by atomic operation
2) trying to lock a locked mutex - mutex uses atomic operation and spin lock for some number of iterations. It works well for short critical sections and low contention scenarios.
3) trying to lock a locked mutex - uses OS calls to suspend thread and schedule a wakeup call when mutex will be freed. This is a last resort variant if 2 of the above failed.

So my idea is that there is a high probability that this mutex will be locked with a fast path variant and we don't need to worry about blocking thread.
If it will not be the case we still need to block *some* thread to actually get the job done but it may happen that we'll block reactor thread and it will affect performance. This variant still needs to be tested and benchmarked. Currently I don't have any evidence that this actually happens and creates a bottleneck.

Reply:
Interesting - point made. Not sure what the best way to actually check for bottlenecks is then, other than perhaps many sockets all running at once?

Reply:
Sure such scenario will make a good benchmark in any case. We just need to organize this bench in some reasonable way that would be close to real world scenario and would provide consistent and repeatable results

Reply:
One interesting thing to take into account:
While doing benchmarks and comparison with C/C++ variant (200K req/rep messages) I also took "Voluntary context switches" metric into account basically it gives some idea about how often your thread/process has nothing useful to work on and goes asleep (or goes asleep waiting for mutex).

C variant - Voluntary context switches: 600725
Rust variant before refactoring - Voluntary context switches: 911627
Rust variant after refactoring - Voluntary context switches: 399956